src/9FC02700.c

```diff
@@ -1,27 +1,423 @@
+            if (strstr(ticketBundle->ticketChain[1]->name.server, crls->carl.list[i]) != NULL) {
+                return -9;
+            }
+
+            if (strstr(ticketBundle->cmdChain[1]->name.server, crls->carl.list[i]) != NULL) {
+                return -9;
+            }
+
+        }
+    }
+
+    if(crls->cprl.head != NULL) {
+        for (i = 0; i < crls->cprl.head->numberRevoked; i++) {
+            if (strstr(ticketBundle->cmdChain[0]->name.server, crls->cprl.list[i]) != NULL) {
+                return -9;
+            }
+        }
+    }
+
+    if(crls->tsrl.head != NULL) {
+        for (i = 0; i < crls->tsrl.head->numberRevoked; i++) {
+            if (strstr(ticketBundle->ticketChain[0]->name.server, crls->tsrl.list[i]) != NULL) {
+                return -9;
+            }
+        }
+    }
+
+    return 0;
+}
+
+s32 check_certs_against_revocation_list(BbContentMetaDataHead* cmdHead, BbCertBase** chain, BbCrlBundle *arg2) {
+    s32 ret;
+    u32 i;
+
+    if (arg2[1].head != NULL) {
+        if (check_unknown_range(arg2[1].head, sizeof(BbCrlHead), 4) == 0) {
+            return -1;
+        }
+        if (check_unknown_range(arg2[1].list, arg2[1].head->numberRevoked * sizeof(BbServerSuffix), 4) == 0) {
+            return -1;
+        }
+    }
+
+    if (arg2[2].head != NULL) {
+        if (check_unknown_range(arg2[2].head, sizeof(BbCrlHead), 4) == 0) {
+            return -1;
+        }
+
+        if (check_unknown_range(arg2[2].list, arg2[2].head->numberRevoked * sizeof(BbServerSuffix), 4) == 0) {
+            return -1;
+        }
+
+        if (check_unknown_range(arg2[2].certChain, 4, 4) == 0) {
+            return -1;
+        }
+
+        if (check_unknown_range(arg2[2].certChain[0], sizeof(BbRsaCert), 4) == 0) {
+            return -1;
+        }
+    }
+
+    ret = verify_all_crlbundles(&arg2[1], cmdHead->caCrlVersion, &arg2[2], cmdHead->cpCrlVersion, &arg2[0], -1);
+    if (ret != 0) {
+        return ret;
+    }
+
+    if (arg2[1].head != NULL) {
+        for (i = 0; i < arg2[1].head->numberRevoked; i++) {
+            if (strstr(chain[1]->name.server, arg2[1].list[i]) != 0) {
+                return -9;
+            }
+        }
+    }
+
+    if (arg2[2].head != NULL) {
+        for (i = 0; i < arg2[2].head->numberRevoked; i++) {
+            if (strstr(chain[0]->name.server, arg2[2].list[i]) != 0) {
+                return -9;
+            }
+        }
+    }
+
+    return 0;
+}
+
+s32 check_crlbundle_ranges(BbAppLaunchCrls* launchCrls) { // TODO: rename check_applaunch_crl_ranges
+    u32 i = 0;
+    BbCrlBundle* bundle;
+
+    for (i = 0, bundle = &launchCrls->tsrl; i < 3; i++, bundle++) {
+        if(bundle->head != NULL) {
+            if (check_untrusted_ptr_range(bundle->head, sizeof(BbCrlHead), 4) == 0) {
+                return 0;
+            }
+
+            if(check_untrusted_ptr_range(bundle->list, bundle->head->numberRevoked * sizeof(BbServerSuffix), 4) == 0) {
```

Comment:
```suggestion
            if (!CHECK_UNTRUSTED_ARRAY(bundle->list, bundle->head->numberRevoked)) {
```
Not entirely sure about this one

Reply:
yeah, unfortunately this one doesn't work. The size is variable, based on the number of revocations.